packages/core/core/src/requests/WriteBundleRequest.js

```diff
@@ -0,0 +1,185 @@
+// @flow strict-local
+
+import type {Async, FilePath} from '@parcel/types';
+import type {StaticRunOpts} from '../RequestTracker';
+import type {Bundle, ContentKey, PackagedBundleInfo} from '../types';
+import type {FileSystem, FileOptions} from '@parcel/fs';
+import type BundleGraph from '../BundleGraph';
+import type {BundleInfo} from '../PackagerRunner';
+
+import {HASH_REF_PREFIX, HASH_REF_REGEX} from '../constants';
+import nullthrows from 'nullthrows';
+import path from 'path';
+import {NamedBundle} from '../public/Bundle';
+import {TapStream} from '@parcel/utils';
+import {Readable, Transform} from 'stream';
+
+const BOUNDARY_LENGTH = HASH_REF_PREFIX.length + 32 - 1;
+
+type WriteBundleRequestInput = {|
+  bundleGraph: BundleGraph,
+  bundle: Bundle,
+  info: BundleInfo,
+  hashRefToNameHash: Map<string, string>,
+|};
+
+type RunInput = {|
+  input: WriteBundleRequestInput,
+  ...StaticRunOpts,
+|};
+
+export type WriteBundleRequest = {|
+  id: ContentKey,
+  +type: 'write_bundle_request',
+  run: RunInput => Async<PackagedBundleInfo>,
+  input: WriteBundleRequestInput,
+|};
+
+/**
+ * Writes a bundle to the dist directory, replacing hash references with the final content hashes.
+ */
+export default function createWriteBundleRequest(
+  input: WriteBundleRequestInput,
+): WriteBundleRequest {
+  let nameHash = nullthrows(
+    input.hashRefToNameHash.get(input.bundle.hashReference),
+  );
+  return {
+    id: `${input.bundle.id}:${input.info.hash}:${nameHash}`,
+    type: 'write_bundle_request',
+    run,
+    input,
+  };
+}
+
+async function run({input, options, api}: RunInput) {
+  let {bundleGraph, bundle, info, hashRefToNameHash} = input;
+  let {inputFS, outputFS} = options;
+  let name = nullthrows(bundle.name);
+  let thisHashReference = bundle.hashReference;
+
+  if (info.type !== bundle.type) {
+    name = name.slice(0, -path.extname(name).length) + '.' + info.type;
+  }
+
+  if (name.includes(thisHashReference)) {
+    let thisNameHash = nullthrows(hashRefToNameHash.get(thisHashReference));
+    name = name.replace(thisHashReference, thisNameHash);
+  }
+
+  let filePath = path.join(bundle.target.distDir, name);
+
+  // Watch the bundle and source map for deletion.
+  // Also watch the dist dir because invalidateOnFileDelete does not currently
+  // invalidate when a parent directory is deleted.
+  // TODO: do we want to also watch for file edits?
```

Comment:
I thought it might be weird if a user wanted to manually edit a file in the dist directory for some reason (e.g. debugging), and Parcel immediately overwrote it